[rush] Rush add causes malformed package.json if the package.json does not have "dependencies" field yet and have >1 entries in "contributors" field

<!-- Have a question?  Before creating an issue, ask in the chat room: https://rushstack.zulipchat.com/ -->

<!-- Try invoking Rush with the "--debug" parameter, for example "rush --debug install".  This prints a full call stack when errors occur. -->

<!--------------------------------------------------------------------------
👉 STEP 1: Write a concise but specific issue title in the box above.
     Example: "[rush] Using --example switch causes TypeError"
--------------------------------------------------------------------------->

## Summary

<!--------------------------------------------------------------------------
👉 STEP 2: In a few sentences, please explain:

     What were you trying to accomplish?
     What action did you perform that ran into trouble?
     What went wrong?
--------------------------------------------------------------------------->


If there are >1 contributors in package.json and there is no dependencies yet, running `rush add --package <pkg-name>` will cause the package.json to be malformed.


## Repro steps

<!--------------------------------------------------------------------------
👉 STEP 3: If your issue is a feature request and not a bug, delete this
     "Repro steps" section and skip to STEP 6.

👉 STEP 4: In many cases we can investigate bugs much faster if you include:
     The URL for a simplified Git branch that reproduces the problem.
     Step by step instructions for how to build the branch and see the error.

👉 STEP 5: It's also helpful to include an "expected" and "actual" result.
     But if that's not relevant, feel free to delete those fields.
--------------------------------------------------------------------------->

https://github.com/kenrick95/rush-repro-rush-add-malformed-package-json


Check `pkg-a/package.json`:

```json
{
  "name": "pkg-a",
  "private": true,
  "version": "0.0.0",
  "contributors": [
    {
      "name": "person-a",
      "email": "example-a@example.org"
    },
    {
      "name": "person-b",
      "email": "example-b@example.org"
    }
  ]
}
```


Repro:

```sh
cd pkg-a
rush add --package meow
```


  **Expected result:** <!-- What you expected these steps to accomplish -->


- A valid JSON in package.json
- rush add runs successfully

```json
{
  "name": "pkg-a",
  "private": true,
  "version": "0.0.0",
  "contributors": [
    {
      "name": "person-a",
      "email": "example-a@example.org"
    },
    {
      "name": "person-b",
      "email": "example-b@example.org"
    }
  ],
  "dependencies": {
    "meow": "~13.2.0"
  }
}
```

  **Actual result:** <!-- If an error occurred, include the full error message text and any call stack. -->


- Malformed JSON in package.json
- rush add failed

```json
{
  "name": "pkg-a",
  "private": true,
  "version": "0.0.0",
  "contributors": [
    {
      "name": "person-a",
      "email": "example-a@example.org"
    },
    {
      "name": "person-b",
      "email": "example-b@example.org"
    }
  ],
  "dependencies": {
    "meow": "~13.2.0",
  }
}
```

```log
➜  pkg-a git:(master) ✗ rush add --package meow
Found configuration in /path/to/repo/rush-repro-rush-add/rush.json


Rush Multi-Project Build Tool 5.124.6 - https://rushjs.io
Node.js version is 18.20.3 (LTS)

Found configuration in /path/to/repo/rush-repro-rush-add/rush.json

Starting "rush add"


Determining new version for dependency: meow
No version selector was specified, so the version will be determined automatically.

Trying to acquire lock for pnpm-8.15.8
Acquired lock for pnpm-8.15.8
Found pnpm version 8.15.8 in /path/to/user/.rush/node-v18.20.3/pnpm-8.15.8

Symlinking "/path/to/repo/rush-repro-rush-add/common/temp/pnpm-local"
  --> "/path/to/user/.rush/node-v18.20.3/pnpm-8.15.8"
The "ensureConsistentVersions" policy is NOT active, so we will assign the latest version.

Querying NPM registry for latest version of "meow"...

Found latest version: 13.2.0

Assigning version "~13.2.0" for "meow".
Updating projects to use meow@~13.2.0

Wrote/path/to/repo/rush-repro-rush-add/pkg-a/package.json

Running "rush update"

Checking Git policy for this repository.

Trying to acquire lock for pnpm-8.15.8
Acquired lock for pnpm-8.15.8
Found pnpm version 8.15.8 in /path/to/user/.rush/node-v18.20.3/pnpm-8.15.8

Symlinking "/path/to/repo/rush-repro-rush-add/common/temp/pnpm-local"
  --> "/path/to/user/.rush/node-v18.20.3/pnpm-8.15.8"
Transforming /path/to/repo/rush-repro-rush-add/common/config/rush/.npmrc
  --> "/path/to/repo/rush-repro-rush-add/common/temp/.npmrc"

Updating workspace files in /path/to/repo/rush-repro-rush-add/common/temp
Copying "/path/to/repo/rush-repro-rush-add/common/config/rush/pnpm-lock.yaml"
  --> "/path/to/repo/rush-repro-rush-add/common/temp/pnpm-lock.yaml"
Copying "/path/to/repo/rush-repro-rush-add/common/config/rush/pnpm-lock.yaml"
  --> "/path/to/repo/rush-repro-rush-add/common/temp/pnpm-lock-preinstall.yaml"

Checking installation in "/path/to/repo/rush-repro-rush-add/common/temp"

Deleting files from /path/to/repo/rush-repro-rush-add/common/temp/node_modules

Running "pnpm install" in /path/to/repo/rush-repro-rush-add/common/temp

 ERR_PNPM_JSON_PARSE  Unexpected token "}" (0x7D) in JSON at position 287 while parsing near "...meow\": \"~13.2.0\",\n  }\n}\n" in /path/to/repo/rush-repro-rush-add/pkg-a/package.json 

  15 |   "dependencies": {
  16 |     "meow": "~13.2.0",
> 17 |   }
     |   ^
  18 | }
  19 |


The command failed:
 /path/to/repo/rush-repro-rush-add/common/temp/pnpm-local/node_modules/.bin/pnpm install --store /path/to/repo/rush-repro-rush-add/common/temp/pnpm-store --config.cacheDir=/path/to/repo/rush-repro-rush-add/common/temp/pnpm-store --config.stateDir=/path/to/repo/rush-repro-rush-add/common/temp/pnpm-store --no-prefer-frozen-lockfile --no-strict-peer-dependencies --config.auto-install-peers=false --config.resolutionMode=highest --config.ignoreCompatibilityDb --recursive --link-workspace-packages false --reporter default
ERROR: Error: The command failed with exit code 1
Giving up after 1 attempts


ERROR: The command failed with exit code 1
```

## Details

<!--------------------------------------------------------------------------
👉 STEP 6: Provide any additional information you think might be helpful:

     What do you think is the cause of this problem?
     How do you think we should fix this?
--------------------------------------------------------------------------->

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | <!-- X.Y.Z --> 5.117.8 |
| `rushVersion` from rush.json? | <!-- X.Y.Z --> 5.129.6 |
| `useWorkspaces` from rush.json? | <!-- X.Y.Z --> Yes |
| Operating system? | <!-- Windows / Mac / Linux --> Mac |
| Would you consider contributing a PR? | <!-- Yes / No --> Yes |
| Node.js version (`node -v`)? | <!-- X.Y.Z --> 18.20.3 |

--- a/common/changes/@microsoft/rush/kenrick-fix-rush-add_2024-06-27-02-39.json
+++ b/common/changes/@microsoft/rush/kenrick-fix-rush-add_2024-06-27-02-39.json
@@ -1,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Fix malformed package.json by explicitly setting jsonSyntax to strict",
+      "type": "none"
+    }
+  ],
+  "packageName": "@microsoft/rush"
+}

--- a/common/changes/@microsoft/rush/kenrick-fix-rush-add_2024-06-27-02-39.json
+++ b/common/changes/@microsoft/rush/kenrick-fix-rush-add_2024-06-27-02-39.json
@@ -2,7 +2,7 @@
   "changes": [
     {
       "packageName": "@microsoft/rush",
-      "comment": "Fix malformed package.json by explicitly setting jsonSyntax to strict",
+      "comment": "Fix an issue where running `rush add` in a project can generate a `package.json` file that uses JSON5 syntax. Package managers expect strict JSON.",
       "type": "none"
     }
   ],

--- a/common/changes/@rushstack/node-core-library/kenrick-fix-rush-add_2024-06-27-02-39.json
+++ b/common/changes/@rushstack/node-core-library/kenrick-fix-rush-add_2024-06-27-02-39.json
@@ -1,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/node-core-library",
+      "comment": "Accept jsonSyntax option at save method to avoid malformed package.json",
+      "type": "patch"
+    }
+  ],
+  "packageName": "@rushstack/node-core-library"
+}

--- a/common/changes/@rushstack/node-core-library/kenrick-fix-rush-add_2024-06-27-02-39.json
+++ b/common/changes/@rushstack/node-core-library/kenrick-fix-rush-add_2024-06-27-02-39.json
@@ -2,8 +2,8 @@
   "changes": [
     {
       "packageName": "@rushstack/node-core-library",
-      "comment": "Accept jsonSyntax option at save method to avoid malformed package.json",
-      "type": "patch"
+      "comment": "Add support for the `jsonSyntax` option to the `JsonFile.save`, `JsonFile.saveAsync`, and `JsonFile.stringify` functions.",
+      "type": "minor"
     }
   ],
   "packageName": "@rushstack/node-core-library"

--- a/common/reviews/api/node-core-library.api.md
+++ b/common/reviews/api/node-core-library.api.md
@@ -416,7 +416,7 @@ export interface IJsonFileSaveOptions extends IJsonFileStringifyOptions {
 }
 
 // @public
-export interface IJsonFileStringifyOptions {
+export interface IJsonFileStringifyOptions extends IJsonFileParseOptions {
     headerComment?: string;
     ignoreUndefinedValues?: boolean;
     newlineConversion?: NewlineKind;

--- a/libraries/node-core-library/src/JsonFile.ts
+++ b/libraries/node-core-library/src/JsonFile.ts
@@ -368,7 +368,7 @@ export class JsonFile {
       });
     } else if (options.prettyFormatting) {
       stringified = jju.stringify(newJsonObject, {
-        mode: explicitMode ? explicitMode : 'json',
+        mode: explicitMode ?? 'json',
         indent: 2
       });
 

--- a/libraries/node-core-library/src/JsonFile.ts
+++ b/libraries/node-core-library/src/JsonFile.ts
@@ -363,7 +363,7 @@ export class JsonFile {
     if (previousJson !== '') {
       // NOTE: We don't use mode=json here because comments aren't allowed by strict JSON
       stringified = jju.update(previousJson, newJsonObject, {
-        mode: explicitMode ? explicitMode : JsonSyntax.Json5,
+        mode: explicitMode ?? JsonSyntax.Json5,
         indent: 2
       });
     } else if (options.prettyFormatting) {

--- a/libraries/node-core-library/src/JsonFile.ts
+++ b/libraries/node-core-library/src/JsonFile.ts
@@ -128,7 +128,7 @@ export interface IJsonFileLoadAndValidateOptions extends IJsonFileParseOptions, 
  *
  * @public
  */
-export interface IJsonFileStringifyOptions {
+export interface IJsonFileStringifyOptions extends IJsonFileParseOptions {
   /**
    * If provided, the specified newline type will be used instead of the default `\r\n`.
    */
@@ -345,17 +345,30 @@ export class JsonFile {
       JsonFile.validateNoUndefinedMembers(newJsonObject);
     }
 
+    let explicitMode: 'json5' | 'json' | 'cjson' | undefined = undefined;
+    switch (options.jsonSyntax) {
+      case JsonSyntax.Strict:
+        explicitMode = 'json';
+        break;
+      case JsonSyntax.JsonWithComments:
+        explicitMode = 'cjson';
+        break;
+      case JsonSyntax.Json5:
+        explicitMode = 'json5';
+        break;
+    }
+
     let stringified: string;
 
     if (previousJson !== '') {
       // NOTE: We don't use mode=json here because comments aren't allowed by strict JSON
       stringified = jju.update(previousJson, newJsonObject, {
-        mode: JsonSyntax.Json5,
+        mode: explicitMode ? explicitMode : JsonSyntax.Json5,
         indent: 2
       });
     } else if (options.prettyFormatting) {
       stringified = jju.stringify(newJsonObject, {
-        mode: 'json',
+        mode: explicitMode ? explicitMode : 'json',
         indent: 2
       });
 

--- a/libraries/rush-lib/src/api/PackageJsonEditor.ts
+++ b/libraries/rush-lib/src/api/PackageJsonEditor.ts
@@ -2,7 +2,7 @@
 // See LICENSE in the project root for license information.
 
 import * as semver from 'semver';
-import { InternalError, type IPackageJson, JsonFile, Sort } from '@rushstack/node-core-library';
+import { InternalError, type IPackageJson, JsonFile, Sort, JsonSyntax } from '@rushstack/node-core-library';
 import { cloneDeep } from '../utilities/objectUtilities';
 
 /**
@@ -308,7 +308,10 @@ export class PackageJsonEditor {
     if (this._modified) {
       this._modified = false;
       this._sourceData = this._normalize(this._sourceData);
-      JsonFile.save(this._sourceData, this.filePath, { updateExistingFile: true });
+      JsonFile.save(this._sourceData, this.filePath, {
+        updateExistingFile: true,
+        jsonSyntax: JsonSyntax.Strict
+      });
       return true;
     }
     return false;

